python/tests/test_tc.py

```diff
@@ -0,0 +1,420 @@
+                    * rsqrt( var(n, g) + 1e-5 )
+                    + beta(g, d)
+            }
+        """
+
+        N, G, D, H, W = 32, 32, 4, 56, 56
+        moments = tc.define(
+            group_normalization,
+            "moments",
+            tuner_config=tuner_config)
+        group_norm = tc.define(
+            group_normalization,
+            "group_normalization",
+            tuner_config=tuner_config)
+
+        I, gamma, beta = (
+            torch.randn(N, G, D, H, W, device='cuda'),
+            torch.randn(G, D, device='cuda'),
+            torch.randn(G, D, device='cuda'))
+
+        moments.tune_and_compile(
+            I.view((N * G, -1)),
+            starting_options=tc.MappingOptions('naive'))
+        mean, var = moments(I.view((N * G, -1)))
+
+        group_norm.tune_and_compile(
+            I,
+            gamma,
+            beta,
+            mean.view((N, G)),
+            var.view((N, G)),
+            starting_options=tc.MappingOptions('naive'))
+        out, = group_norm(I, gamma, beta, mean.view((N, G)), var.view((N, G)))
+
+        # With the previously compiled TCs we can just create a forward-only function
+        def forward(I, gamma, beta):
+            mean, var = moments(I.view((N * G, -1)))
+            out, = group_norm(I, gamma, beta, mean.view((N, G)), var.view((N, G)))
+            return out
+
+        # don't call this group_norm or it will change the captured variable in forward
+        group_norm_function = tc.define_with_autograd(forward)
+        out = group_norm_function(I, gamma, beta)
+        # TODO: test results!!!
+
+    #
+    # This tests single kernel forward/backward with tc.define_with_autograd.
+    #
+    def test_conv_backward_fused(self):
+        conv = """
+        def convolution(float(N,C,H,W) I, float(M,C,KH,KW) W1) -> (O) {
+            O(n, m, h, w) +=!
+                I(n, r_c, h + r_kh, w + r_kw) * W1(m, r_c, r_kh, r_kw)
+        }
+        def convolution_grad(
+            float(N,C,H,W) I, float(M,C,KH,KW) W1, float(N,M,H,W) d_O)
+            -> (d_I, d_W1)
+        {
+            d_I(n, c, h, w) +=!
+                d_O(  n, r_m, h - r_kh, w - r_kw) * W1(r_m, c, r_kh, r_kw)
+            d_W1(m, c, kh, kw) +=!
+                d_O(r_n,   m, r_h - kh, r_w - kw) *  I(r_n, c,  r_h,  r_w)
+        }
+        """
+
+        N, C, H, W, O, kH, kW = 32, 4, 56, 56, 16, 1, 1
+        convolution = tc.define(
+            conv,
+            "convolution",
+            tuner_config=tuner_config)
+        convolution_grad = tc.define(
+            conv,
+            "convolution_grad",
+            tuner_config=tuner_config)
+        I, W = (
+            torch.randn(N, C, H, W, device='cuda', requires_grad=True),
+            torch.randn(O, C, kH, kW, device='cuda', requires_grad=True))
+
+        convolution_function = tc.define_with_autograd(
+            lambda I, W: convolution(I, W),
+            lambda I, W, d_O: convolution_grad(I, W, d_O)
```

Comment:
nit: you can eta-reduce those to `tc.define_with_autograd(convolution, convolution_grad)`